Weird bug I cant figure out: permission denied

Reported simultaneously today:

- https://github.com/EnvironmentOntology/envo/pull/1330#issuecomment-1210543852
- https://github.com/Southern-Cross-Plant-Science/cdno/issues/97#issuecomment-1210570928

https://github.com/Southern-Cross-Plant-Science/cdno/blob/master/src/ontology/Makefile#L114

I cant figure out what is wrong. Here is what I know:

If I run using the ODK docker wrapper:

```
./run.sh dosdp-tools generate --table-format=csv --template=./patterns/entity_attribute.yaml --outfile=modules/entity_attribute.tmp.owl --obo-prefixes=true --infile=modules/entity_attribute.csv
```

Everything works fine.

While adding the absolute identical command into a Makefile:

```
ttt:
	dosdp-tools generate --table-format=csv --template=./patterns/entity_attribute.yaml --outfile=modules/entity_attribute.tmp.owl --obo-prefixes=true --infile=modules/entity_attribute.csv
```

Does not work:

```
make: dosdp-tools: Permission denied
make: *** [Makefile:66: ttt] Error 127
```

Strangely enough, the makefile works without the docker wrapper.

Any help appreciated.

For information, contrary to what many of us (myself included) thought while we were debugging this issue yesterday on Slack, this is _not_ a Docker-induced problem.

I can reproduce the problem without Docker on my GNU/Linux system:

1) Create a directory hierarchy somewhere (here, my home folder, but it doesn’t matter) with a `dosdp-tools` in the middle:

```sh
make -p ~/tools/dosdp-tools/bin
```

2) Create an executable file named `dosdp-tools` at the bottom of this hierarchy:

```sh
cat <<EOF > ~/tools/dosdp-tools/bin/dosdp-tools
#!/bin/sh
echo "I am here"
EOF
chmod 755 ~/tools/dosdp-tools/bin/dosdp-tools
```

3) Create a Makefile with a recipe that tries to call the `dosdp-tools` script:

```sh
cat <<EOF > ~/Makefile
test:
        @dosdp-tools
EOF
```

4) Try invoking that recipe with a custom PATH that includes both the `~/tools` directory _and_ the `~/tools/dosdp-tools/bin` directory (in that order):

```sh
PATH=$HOME/tools:$HOME/tools/dosdp-tools/bin:$PATH make test
make: dosdp-tools: Permission denied
make: *** [Makefile:2: test] Error 127
```

5) Observe that if you invert the entries in the custom PATH so that the `~/tools/dosdp-tools/bin` folder is _before_ the `~/tools` folder, everything is fine:

```sh
PATH=$HOME/tools/dosdp-tools/bin:$HOME/tools:$PATH make test
I am here
```

So there is no Docker trickery involved here. The problem lies purely within Make, which seems to have its own logic to find an executable in the PATH (instead of using, say, `execvp(3)`), and that logic is confused when it finds in the PATH a _directory_ with the same name than the executable it is looking for.

Reproduced with GNU Make version 4.3 on GNU/Linux (which is also the version used in the ODK since version 1.3.1); _not_ reproduced with GNU Make 3.81 on MacOS, and _not_ reproduced with GNU Make 4.2.1 (which is the version used in the ODK up to version 1.3.0, before we upgraded to Ubuntu 22.04).

So, looks like a bug introduced in GNU Make between 4.2.1 and 4.3. I guess people who were not affected by the problem yesterday were still using a pre-1.3.1 ODK.

This also highlight a lack of coverage of our tests. We did a lot of testing after upgrading the base system to Ubuntu 22.04, but we completely missed this problem.

--- a/docker/odklite/Dockerfile
+++ b/docker/odklite/Dockerfile
@@ -10,7 +10,7 @@ WORKDIR /tools
 ENV JAVA_HOME="/usr"
 ENV LC_ALL=C.UTF-8
 ENV LANG=C.UTF-8
-ENV PATH="/tools:/tools/dosdp-tools/bin:$PATH"
+ENV PATH="/tools:/tools/dosdptools/bin:$PATH"
 
 ARG ODK_VERSION=0.0.0
 ENV ODK_VERSION=$ODK_VERSION
@@ -47,7 +47,7 @@ RUN wget -nv $ROBOT_JAR \
 RUN wget -nv https://github.com/INCATools/dosdp-tools/releases/download/v$DOSDP_VERSION/dosdp-tools-$DOSDP_VERSION.tgz && \
     tar zxf dosdp-tools-$DOSDP_VERSION.tgz && \
     rm dosdp-tools-$DOSDP_VERSION.tgz && \
-    mv dosdp-tools-$DOSDP_VERSION /tools/dosdp-tools && \
+    mv dosdp-tools-$DOSDP_VERSION /tools/dosdptools && \
     wget -nv --no-check-certificate https://raw.githubusercontent.com/INCATools/dead_simple_owl_design_patterns/master/src/simple_pattern_tester.py \
         -O /tools/simple_pattern_tester.py && \
     chmod 755 /tools/simple_pattern_tester.py

